packages/data-service/src/user-data.ts

```diff
@@ -0,0 +1,289 @@
+ *      darkMode: {type: 'boolean'}
+ *    }
+ * };
+ *
+ * const ajv = new Ajv();
+ * const serialize = ajv.compileSerializer(schema);
+ * const parse = ajv.compileParser(schema);
+ *
+ * const userData = new UserData<UserPreferences>({serialize, parse});
+ *
+ */
+export class UserData<T = unknown> {
+  private _basePath: string;
+  private _parse: ThrowingParseFunction<T>;
+  private _serialize: SerializeFunction<T>;
+
+  constructor(
+    options: {
+      /**
+       * Alternate base path instead of electronApp.getPath('userData')
+       * useful for testing and non electron environments.
+       *
+       * Default to `electronApp.getPath('userData')`
+       */
+      alternateDataRoot?: string;
+
+      /**
+       * Subpath of the user data dir to use as base path for any operation.
+       *
+       * Default to `./`
+       */
+      subDir?: string;
+
+      /**
+       * An optional parse function which parses a string to an object of the
+       * generic type `T`.
+       */
+      parse?: ParseFunction<T>;
+
+      /**
+       * An optional serialize function that converts an object of the
+       * generic type `T` to a string.
+       */
+      serialize?: SerializeFunction<T>;
+    } = {}
+  ) {
+    const electronApp = electron.remote?.app || electron.app;
+    const userDataRoot = path.resolve(
+      options.alternateDataRoot || electronApp.getPath('userData')
+    );
+
+    this._basePath = options.subDir
+      ? this._resolve(userDataRoot, options.subDir)
+      : userDataRoot;
+
+    this._parse = this._conformParseFunctionToThrowOnError(
+      options.parse ?? this._defaultParser
+    );
+
+    this._serialize = options.serialize ?? this._defaultSerializer;
+  }
+
+  /**
+   * Reads and parse all the files in the configured subdir of user
+   * data that matches the glob pattern `pattern`.
+   *
+   * NOTE: this method will log in case of errors and try to load as much files
+   * as possible.
+   *
+   * @param {string} [pattern = "*.*"] - A glob pattern matching all the files to load.
+   * @returns {Promise<T[]>} all the files read.
+   */
+  async readAll(pattern = '*.*'): Promise<T[]> {
+    const pathAndContents = await this._readAllWithFileNames(pattern);
+
+    const results = [];
+
+    for (const { path, content } of pathAndContents) {
+      try {
+        results.push(JSON.parse(content.toString()));
+      } catch (error) {
+        log.error(
+          mongoLogId(1_001_000_061),
+          'UserData',
+          'Error parsing file to json',
+          {
+            path: path,
+          }
+        );
+      }
+    }
+
+    return results;
+  }
+
+  /**
+   * Reads and parse the file specified by `filePath`.
+   *
+   * NOTE: this function will throw in case of error.
+   *
+   * @param {string} filePath - the file path to read.
+   * @returns {Promise<T>} the parsed content.
+   */
+  async read(filePath: string): Promise<T> {
+    const content = (await this._readFile(filePath)).toString();
```

Comment:
Just a note, but using `fs.readFile(path, 'utf8')` instead if `fs.readFile(path).toString()` is generally a bit preferable because it skips an extra step + avoids copying the file around … doesn’t matter here and it’s fine if you leave it as is, but it always looks a bit odd to me :)

So since we’re currently always reading UTF-8 text files, and the parse/serialize type definitions here also refer to strings, I think it’s okay to just always directly read the data as strings.

Reply:
Yeah makes lots of sense!